students/brian_warn/session06/test_mailroom.py

```diff
@@ -0,0 +1,18 @@
+#!/Library/Frameworks/Python.framework/Versions/3.6/bin/python
+ # test_mailroom.py
+
+import mailroom_from_chb as mr
+
+import pytest
+
+
+def test_get_donor_db():
+    result = mr.get_donor_db()
+    assert result == {'william gates iii': ("William Gates III", [653772.32, 12.17]),
```

Comment:
Careful with tests like this. In a way, you are not testing the logic in the function, but rather the data that's there. And that is just a test data set. If this code were moved into production, that data would not stay the same.

I think I chose not to test it at all -- if the rest of the tests pass, then there must be a valid donor_db object.

This actually gave me an idea. the get_donor_db function really is simply supplying a test dataset.

We needed it, so that we could test and demonstrate the code. But now that we're adding tests, it's probably better that the test data be embedded in the tests!

So you may want to copy the contents of the get_donor_db function into your tests:

```
# so that it's there for the tests
mailroom.donor_db = {'william gates iii': ("William Gates III", [653772.32, 12.17]),
                     'jeff bezos': ("Jeff Bezos", [877.33]),
                     'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
                     'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
                     }
```
Now you have a test dataset to use, AND most importantly, you've defined the test dataset in the tests themselves - which is key, as you need to use the values in your tests.

And now you can change the sample data in mailroom itself, or change where it gets it, etc, without breaking your tests.

Reply:
Thank you @PythonCHB (cc'ing @codedragon) for this helpful info.  This does indeed work -- assuming I'm implementing this test correctly.  Is the test as I've now defined it (below) the right way to perform this test?
```
def test_get_donor_db():
    mr.donor_db = {'william gates iii': ("William Gates III", [653772.32, 12.17]),
                     'jeff bezos': ("Jeff Bezos", [877.33]),
                     'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
                     'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
                     }
    result = mr.get_donor_db()
    # assert result == {'william gates iii': ("William Gates III", [653772.32, 12.17]),
    #     'jeff bezos': ("Jeff Bezos", [877.33]),
    #     'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
    #     'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),}
    assert result == mr.donor_db
```

Reply:
actually, not quite :-)

in the line:
`mr.donor_db = {'william ga .... `

you are setting the donor_db name in the mailroom module.

then in the next lines:

```
result = mr.get_donor_db()`
assert result == mr.donor_db
```
You are checking that get_donor_db returns the same data.

But: you don't need to set `mr.get_donor_db` to test if you are getting what you want from `get_donor_db`. And youshoudn't, because that's a side effect of the test that you don't want -- the mailroom module has now been modified by this test, even though there was no need to.

So you could just change `mr.donr_db` to `donor_db` in two places, as you're good. 

Then there is the question of whether this is something to test.

mr.get_donor_db is providing a base donor database for the rest of the code to work with.

At this point, it's a hard-coded sample database. But if this code were ever deployed -- that sample data would be replaced with real data -- or an empty dataset, or...

so testing that the sample data is what you expect isn't really useful.

maybe I should have called the function: get_sample_donor_data() :-)

If you do test get_donor_db -- I'd test that you got a valid database -- it should be a dict, and if it isn't empty, the keys in the dict should be strings, and the values tuples with the first element a string, and the second a list of numbers.

But I'm not sure that's worth doing -- you want to test actual functionality -- and loading a big literal is expected to work :-)

And if the sample data is bad -- all the other tests are going to fail :-)

Reply:
Thank you @PythonCHB for this response since I now see that I need to answer the simple "Should I test this function?" before setting up a test in the first place.  I had been proceeding with the approach that every function does require unit testing.

Reply:
Yup! The pedantic approach is that every function needs testing -- but in some cases it may not help :-)

I do sometimes put "do nothing" tests it, 'cause then you are at least teseting that the function exists. So no harm in doing:

```
def test_get_donor_db()
    assert mr.get_donor_db is not None
```
it does test that the function exists, and that it does return _something_.

And it'll get your test coverage numbers up :-)

Reply:
oops, jsut saw a typo in that -- the do almost nothing test should be:

```
def test_get_donor_db()
    assert mr.get_donor_db() is not None
```

note that it is CALLING get_donor_db -- assuring you that it can be run :-)